### PR TITLE
Debt/met 5444 Optimization of validation workflow

### PR DESCRIPTION
**Pull request by JoanaCMS**

### CODE REVIEW

File: src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractor.java [hunk 1]
```diff
@@ -7,20 +7,22 @@
...
-import java.util.ArrayList;
-import java.util.List;
 import java.util.stream.Collectors;
 
 /**
  * The type Validated record extractor.
  */
-public class ValidatedRecordExtractor implements ValidationExtractor {
+public class ValidatedRecordExtractor {
     private static final Logger LOGGER = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractor.java:17]: Done!

File: src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractor.java [hunk 2]
```diff
@@ -32,21 +34,25 @@ public Record extractRecord(RecordInfo validatedRecordInfo) {
...
-                                                 List<ValidationResult> validationResults) {
-        List<ValidationResult> result = new ArrayList<>(validationResults);
+    /**
+     * Extract results list.
+     *
+     * @param step              the step
+     * @param recordInfo        the record info
+     * @return the list
+     */
+    public static ValidationStepContent extractResults(Step step,RecordInfo recordInfo) {
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractor.java:44]: Renamed it

File: src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/TransformationValidationStep.java [hunk 3]
```diff
@@ -11,54 +11,40 @@
...
     @Override
-    public void setNextValidationStep(ValidationStep nextValidationStep) {
-        this.nextValidationStep = nextValidationStep;
-    }
-
-    @Override
-    public List<ValidationResult> performStep(Record recordToValidate) {
-        List<ValidationResult> validationResults = new ArrayList<>();
+    public ValidationStepContent performStep(Record recordToValidate) {
+        ValidationStepContent validationResult;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/TransformationValidationStep.java:36]: Done!

File: src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidationWorkflowService.java [hunk 4]
```diff
@@ -143,4 +145,34 @@ private void finalizeValidationPatternAnalysis(String datasetId, ExecutionPoint
             LOGGER.debug("Finalize record analysis: {} lock, Unlocked", datasetId);
         }
     }
+
+    private List<ValidationResult> performSteps(Record recordToProcess){
+        List<ValidationResult> validationResults = new ArrayList<>();
+
+        ValidationStepContent harvestResult = harvestValidationStep.performStep(recordToProcess);
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidationWorkflowService.java:152]: Renamed it

File: src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidationWorkflowService.java [hunk 5]
```diff
@@ -143,4 +145,34 @@ private void finalizeValidationPatternAnalysis(String datasetId, ExecutionPoint
...
+        }
+
+        ValidationStepContent transformationResult = transformationValidationStep.performStep(externalValidationResult.getRecordStepResult());
+        validationResults.add(transformationResult.getValidationStepResult());
+
+        if(transformationResult.getValidationStepResult().getStatus() == ValidationResult.Status.FAILED){
+            return validationResults;
+        }
+
+        ValidationStepContent internalValidationResult = internalValidationValidationStep.performStep(transformationResult.getRecordStepResult());
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/main/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidationWorkflowService.java:173]: done

File: src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/HarvestValidationStepTest.java [hunk 6]
```diff
@@ -12,25 +12,19 @@
...
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.doNothing;
-import static org.mockito.Mockito.doThrow;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/HarvestValidationStepTest.java:21]: Fixed

File: src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/HarvestValidationStepTest.java [hunk 7]
```diff
@@ -48,19 +42,15 @@ void validate_expectSuccess() {
...
                 .content("info".getBytes(StandardCharsets.UTF_8))
                 .build();
-        when(externalValidationStep.performStep(any())).thenReturn(List.of(new ValidationResult(Step.VALIDATE_EXTERNAL,
-                new RecordValidationMessage(RecordValidationMessage.Type.INFO, "success"),
-                ValidationResult.Status.PASSED)));
-        harvestValidationStep.setNextValidationStep(externalValidationStep);
 
         //when
-        List<ValidationResult> validationResults = harvestValidationStep.performStep(record);
+        ValidationStepContent validationResults = harvestValidationStep.performStep(record);
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/HarvestValidationStepTest.java:47]: Done

File: src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/InternalValidationValidationStepTest.java [hunk 8]
```diff
@@ -14,11 +14,9 @@
 import org.mockito.junit.jupiter.MockitoExtension;
 
 import java.nio.charset.StandardCharsets;
-import java.util.List;
 import java.util.Optional;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.*;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/InternalValidationValidationStepTest.java:19]: Fixed

File: src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/TransformationValidationStepTest.java [hunk 9]
```diff
@@ -14,11 +14,9 @@
 import org.mockito.junit.jupiter.MockitoExtension;
 
 import java.nio.charset.StandardCharsets;
-import java.util.List;
 import java.util.Optional;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.*;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/TransformationValidationStepTest.java:19]: Fixed

File: src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractorTest.java [hunk 10]
```diff
@@ -9,15 +9,12 @@
 import org.junit.jupiter.api.Test;
 
 import java.nio.charset.StandardCharsets;
-import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
 
-import static org.junit.jupiter.api.Assertions.assertArrayEquals;
-import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.*;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractorTest.java:15]: Fixed

File: src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractorTest.java [hunk 11]
```diff
@@ -63,13 +60,13 @@ void extractResults_expectSuccess() {
                 .build();
         RecordInfo recordInfo = new RecordInfo(expectedRecord);
         // when
-        List<ValidationResult> validationResults = validatedRecordExtractor.extractResults(Step.HARVEST_FILE, recordInfo, new ArrayList<>());
+        ValidationStepContent validationResults = ValidatedRecordExtractor.extractResults(Step.HARVEST_FILE, recordInfo);
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractorTest.java:63]: Done

File: src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractorTest.java [hunk 12]
```diff
@@ -89,13 +86,12 @@ void extractResults_withErrors_expectSuccess() {
         RecordInfo recordInfo = new RecordInfo(expectedRecord, List.of(new RecordError("Fail message1", "stackTrace1"),
                 new RecordError("Fail message2", "stackTrace2")));
         // when
-        List<ValidationResult> validationResults = validatedRecordExtractor.extractResults(Step.HARVEST_FILE, recordInfo, new ArrayList<>());
+        ValidationStepContent validationResults = ValidatedRecordExtractor.extractResults(Step.HARVEST_FILE, recordInfo);
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply src/test/java/eu/europeana/metis/sandbox/service/validationworkflow/ValidatedRecordExtractorTest.java:89]: Done

**Review by jeortizquan — APPROVED** [review]

**Pull request merged**